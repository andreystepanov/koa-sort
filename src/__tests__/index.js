import order from '../index'

const getContext = (request = {}, state = {}) => ({
  request: {
    method: 'GET',
    query: {},
    ...request,
  },
  state,
})
const next = jest.fn()
const middleware = order({ stateParam: 'order' })

beforeEach(() => {
  next.mockClear()
})

test('order middleware is defined', () => {
  expect(order).toBeDefined()
  expect(typeof order).toBe('function')
  expect(middleware).toBeDefined()
  expect(typeof middleware).toBe('function')
})

test("ignores sort property if it's not a GET request", () => {
  const context = getContext({ method: 'POST' })
  middleware(context, next)

  expect(context.state).toEqual({})
  expect(next).toBeCalledTimes(1)
})

test('ignores empty sort', () => {
  const context = getContext()
  middleware(context, next)

  expect(context.state).toEqual({})
})

test('sets single "created_at desc"', () => {
  const context = getContext({ query: { sort: 'created_at' } })
  middleware(context, next)

  expect(context.state).toMatchSnapshot()
})

test('sets single "created_at asc"', () => {
  const context = getContext({ query: { sort: '-created_at' } })
  middleware(context, next)

  expect(context.state).toMatchSnapshot()
})

test('sets multiple order params', () => {
  const context = getContext({ query: { sort: 'active,-created_at' } })
  middleware(context, next)

  expect(context.state).toMatchSnapshot()
})

test('returns state.order property as object', () => {
  const middlewareWithOptions = order({ asObject: true, stateParam: 'order' })
  const context = getContext({
    query: { sort: 'active,-created_at' },
  })

  middlewareWithOptions(context, next)

  expect(context.state).toMatchSnapshot()
})

test('reads from custom query param and sets it to custom state param', () => {
  const middlewareWithOptions = order({ param: 'order', stateParam: 'custom' })
  const context = getContext({
    query: { order: 'active,-created_at' },
  })

  middlewareWithOptions(context, next)

  expect(context.state).toMatchSnapshot()
})
