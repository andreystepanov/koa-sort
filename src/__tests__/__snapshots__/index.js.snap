// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reads from custom query param and sets it to custom state param 1`] = `
Object {
  "custom": Array [
    Array [
      "active",
      "desc",
    ],
    Array [
      "created_at",
      "asc",
    ],
  ],
}
`;

exports[`returns state.order property as object 1`] = `
Object {
  "order": Object {
    "active": "desc",
    "created_at": "asc",
  },
}
`;

exports[`sets multiple order params 1`] = `
Object {
  "order": Array [
    Array [
      "active",
      "desc",
    ],
    Array [
      "created_at",
      "asc",
    ],
  ],
}
`;

exports[`sets single "created_at asc" 1`] = `
Object {
  "order": Array [
    Array [
      "created_at",
      "asc",
    ],
  ],
}
`;

exports[`sets single "created_at desc" 1`] = `
Object {
  "order": Array [
    Array [
      "created_at",
      "desc",
    ],
  ],
}
`;
